_posts/2021-05-27-Chatbot-Immo.md

```diff
@@ -0,0 +1,711 @@
+    GoogleClientSecrets.Details web = new GoogleClientSecrets.Details();
+    web.setClientId(googleApiClient);
+    web.setClientSecret(googleApiSecret);
+    clientSecrets = new GoogleClientSecrets().setWeb(web);
+    httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    flow = new GoogleAuthorizationCodeFlow.Builder(httpTransport, JSON_FACTORY, clientSecrets,
+            Collections.singleton(CalendarScopes.CALENDAR)).build();
+    authorizationUrl = flow.newAuthorizationUrl()
+            .setRedirectUri(redirectUri)
+            .setAccessType("offline");
+    System.out.println("cal authorizationUrl->" + authorizationUrl);
+    return authorizationUrl.build();
+}
+```
+
+In the code above we get the url where the user should login. 
+We need to pass the client id and the client secret with the request as well as the redirect URI. 
+You could also see that I only ask for the calendar scope, which is the only one I needed for this project. 
+When the user opens this URL in the browser, the user has to login to his/her Google account after which Google prompts the user for consent to grant access to the calendar. 
+When the login is successful, the server will send the authorization code back to the redirect URI specified in the request.
+
+#### Exchange authorization code for refresh and access tokens
+When the redirect URI is called, the following function is executed:
+
+```Java
+@Override
+public ResponseEntity<String> callback(String code) {
+    com.google.api.services.calendar.model.Events eventList;
+    System.out.println(code);
+    String message;
+    try {
+        TokenResponse response = flow.newTokenRequest(code).setRedirectUri(redirectUri).execute();
+        credential = flow.createAndStoreCredential(response, "userID");
+        System.out.println(credential.getAccessToken() + "----" + credential.getRefreshToken());
+        client = new com.google.api.services.calendar.Calendar.Builder(httpTransport, JSON_FACTORY, credential).build();
+        Events events = client.events();
+        eventList = events.list("primary").setTimeMin(date1).setTimeMax(date2).execute();
+        message = eventList.getItems().toString();
+        System.out.println("My:" + eventList.getItems());
+    } catch (Exception e) {
+        LOG.warn("Exception while handling OAuth2 callback (" + e.getMessage() + ")."
+                + " Redirecting to google connection status page.");
+        message = "Exception while handling OAuth2 callback (" + e.getMessage() + ")."
+                + " Redirecting to google connection status page.";
+    }
+
+    System.out.println("cal message:" + message);
+    return new ResponseEntity<>(message, HttpStatus.OK);
+}
+```
+
+This is the part where we are exchanging the authorization code for the access and refresh tokens.
+When I got those tokens from the server, I saved the refresh token in the application.properties file so I could use it later to get new access tokens.
```

Comment:
You can use backticks for inline code styling. Or to refer to files.
```suggestion
When I got those tokens from the server, I saved the refresh token in the `application.properties` file so I could use it later to get new access tokens.
```

Reply:
I don't really understand these part.
The `application.properties` file contains the config for the application. How is something being saved to it at runtime?